# Creating a simple Neural Network for Hand Wriiten Number Recognition

Importing requisite Libraries

In [128]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow import keras


# Importing Data

In [129]:
train = pd.read_csv("mnist_train.csv")
eval = pd.read_csv("mnist_test.csv")

In [130]:
train.head()

label  1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  ...  28x19  28x20  \
0      5    0    0    0    0    0    0    0    0    0  ...      0      0   
1      0    0    0    0    0    0    0    0    0    0  ...      0      0   
2      4    0    0    0    0    0    0    0    0    0  ...      0      0   
3      1    0    0    0    0    0    0    0    0    0  ...      0      0   
4      9    0    0    0    0    0    0    0    0    0  ...      0      0   

   28x21  28x22  28x23  28x24  28x25  28x26  28x27  28x28  
0      0      0      0      0      0      0      0      0  
1      0      0      0      0      0      0      0      0  
2      0      0      0      0      0      0      0      0  
3      0      0      0      0      0      0      0      0  
4      0      0      0      0      0      0      0      0  

[5 rows x 785 columns]

Seperating Labels and Features

In [131]:
train_label = train.pop("label")
eval_label = eval.pop("label")

print(train_label)

0        5
1        0
2        4
3        1
4        9
        ..
59995    8
59996    3
59997    5
59998    6
59999    8
Name: label, Length: 60000, dtype: int64


In [132]:
np.reshape(train.loc[0].to_numpy(),(28,28))

array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   3,
         18,  18,  18, 126, 136, 175,  26, 166, 255, 247, 127,   0,   0,
          0,   0],
       [  

Normalizing the data

In [133]:
train = train.div(255)

# Making and Training the Model

In [134]:
model = keras.Sequential([keras.layers.Dense(10, input_shape=(784,), activation='sigmoid')])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy',metrics=['accuracy'])

model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.4667 - accuracy: 0.8783
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3033 - accuracy: 0.9151
Epoch 3/5
1875/1875 [==============================] - 2s 982us/step - loss: 0.2834 - accuracy: 0.9216
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2728 - accuracy: 0.9235
Epoch 5/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2667 - accuracy: 0.9261


Evaluating the accuracy

In [135]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 2ms/step - loss: 47.4034 - accuracy: 0.9141


[47.403419494628906, 0.9140999913215637]

In [136]:
label_predicted = model.predict(eval.loc[:])
print("predicted: ",np.argmax(label_predicted[4]))
print("Actual: ",eval_label[4])

313/313 [==============================] - 0s 1ms/step
predicted:  4
Actual:  4


# Adding Hidden layers

And testing different activation functions.

In [137]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='sigmoid'),
    keras.layers.Dense(10, activation='sigmoid')
    ])


In [138]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4289 - accuracy: 0.8900
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2050 - accuracy: 0.9415
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1522 - accuracy: 0.9564
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1205 - accuracy: 0.9656
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0987 - accuracy: 0.9717


In [139]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 2ms/step - loss: 0.1125 - accuracy: 0.9658


[0.1125175878405571, 0.9657999873161316]

In [140]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='relu'),
    keras.layers.Dense(10, activation='sigmoid')
    ])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2776 - accuracy: 0.9211
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1285 - accuracy: 0.9625
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0886 - accuracy: 0.9736
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0679 - accuracy: 0.9788
Epoch 5/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.0533 - accuracy: 0.9832


In [141]:
model.evaluate(eval.loc[:],eval_label)


313/313 [==============================] - 1s 1ms/step - loss: 15.3923 - accuracy: 0.9739


[15.392298698425293, 0.9739000201225281]

In [142]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='tanh'),
    keras.layers.Dense(10, activation='sigmoid')
    ])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2879 - accuracy: 0.9175
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1421 - accuracy: 0.9589
Epoch 3/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.0999 - accuracy: 0.9713
Epoch 4/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.0753 - accuracy: 0.9781
Epoch 5/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.0590 - accuracy: 0.9835


In [143]:
model.evaluate(eval.loc[:],eval_label)


313/313 [==============================] - 0s 1ms/step - loss: 0.1082 - accuracy: 0.9659


[0.10817556083202362, 0.9659000039100647]

# Testing different optimisers

Testing RMSprop

In [144]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='relu'),
    keras.layers.Dense(10, activation='sigmoid')
    ])

model.compile(optimizer='RMSprop', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2739 - accuracy: 0.9217
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1318 - accuracy: 0.9616
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0973 - accuracy: 0.9718
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0785 - accuracy: 0.9775
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0674 - accuracy: 0.9804


In [145]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 1ms/step - loss: 16.8595 - accuracy: 0.9746


[16.859472274780273, 0.9746000170707703]

Testing SGD

In [146]:
model.compile(optimizer='SGD', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0442 - accuracy: 0.9885
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0407 - accuracy: 0.9897
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0391 - accuracy: 0.9899
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0379 - accuracy: 0.9902
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0369 - accuracy: 0.9907


In [147]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 2ms/step - loss: 14.5689 - accuracy: 0.9768


[14.568870544433594, 0.9768000245094299]

# Adding drop-out Layers

Adding drop out **0.3**

In [148]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(10, activation='sigmoid')
    ])

model.compile(optimizer='SGD', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.7655 - accuracy: 0.7825
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.4234 - accuracy: 0.8783
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3558 - accuracy: 0.8987
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3169 - accuracy: 0.9086
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2893 - accuracy: 0.9172


In [149]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 1ms/step - loss: 26.6025 - accuracy: 0.9349


[26.60251235961914, 0.9348999857902527]

Adding Dropout **0.5**

In [150]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10, activation='sigmoid')
    ])

model.compile(optimizer='SGD', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.8890 - accuracy: 0.7334
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.4909 - accuracy: 0.8584
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.4150 - accuracy: 0.8817
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3707 - accuracy: 0.8934
Epoch 5/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3418 - accuracy: 0.9018


In [151]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 2ms/step - loss: 26.2630 - accuracy: 0.9344


[26.26298713684082, 0.9344000220298767]

Adding Dropout **0.8**

In [152]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='relu'),
    keras.layers.Dropout(0.8),
    keras.layers.Dense(10, activation='sigmoid')
    ])

model.compile(optimizer='SGD', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 1.2749 - accuracy: 0.5752
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.8267 - accuracy: 0.7334
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.7189 - accuracy: 0.7725
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.6660 - accuracy: 0.7896
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.6241 - accuracy: 0.8028


In [153]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 2ms/step - loss: 29.6043 - accuracy: 0.9191


[29.604333877563477, 0.9190999865531921]

# Trying different Loss functions


Sparse_categorical_crossentropy

In [161]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='relu'),
    keras.layers.Dense(10, activation='sigmoid')
    ])

model.compile(optimizer='SGD', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.6631 - accuracy: 0.8341
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3391 - accuracy: 0.9054
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2911 - accuracy: 0.9185
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2621 - accuracy: 0.9263
Epoch 5/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2402 - accuracy: 0.9327


In [155]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 2ms/step - loss: 33.0489 - accuracy: 0.9269


[33.048927307128906, 0.9269000291824341]

mean_absolute_error

In [156]:
model.compile(optimizer='SGD', loss='mean_absolute_error',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 5s 2ms/step - loss: 3.7400 - accuracy: 0.8875
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 3.6450 - accuracy: 0.8133
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 3.6408 - accuracy: 0.7787
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 3.6388 - accuracy: 0.7573
Epoch 5/5
1875/1875 [==============================] - 5s 2ms/step - loss: 3.6361 - accuracy: 0.7397


In [157]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 2ms/step - loss: 3.6192 - accuracy: 0.0980


[3.6192028522491455, 0.09799999743700027]

mean_squared_error

In [158]:
model.compile(optimizer='SGD', loss='mean_squared_error',metrics=['accuracy'])
model.fit(train.loc[:], train_label, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 20.2671 - accuracy: 0.6995
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 20.2631 - accuracy: 0.6529
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 20.2614 - accuracy: 0.6204
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 20.2603 - accuracy: 0.5924
Epoch 5/5
1875/1875 [==============================] - 3s 1ms/step - loss: 20.2594 - accuracy: 0.5688


# Adding early stopping

In [159]:
model = keras.Sequential([
    keras.layers.Dense(100, input_shape=(784,), activation='relu'),
    keras.layers.Dense(10, activation='sigmoid')
    ])

model.compile(optimizer='SGD', loss='sparse_categorical_crossentropy',metrics=['accuracy'])

early_stopping = keras.callbacks.EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)

model.fit(train.loc[:], train_label, epochs=5,callbacks=[early_stopping])



Epoch 1/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.6633 - accuracy: 0.8327
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3368 - accuracy: 0.9071
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2859 - accuracy: 0.9206
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2551 - accuracy: 0.9289
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2324 - accuracy: 0.9352


In [160]:
model.evaluate(eval.loc[:],eval_label)

313/313 [==============================] - 1s 2ms/step - loss: 28.0101 - accuracy: 0.9386


[28.010116577148438, 0.9386000037193298]

# Inferences

1. Base accuracy of the model using one single layer, ***sigmoid*** as the activation function and ***adam*** as the optimiser has about 91% accuracy.

2. Adding a ***Relu*** activated layer with a hundred output node as the first layer and then a ***sigmoid*** hidden layer imporves the accuarcy to about 97%.

3. Setting the optimiser to ***SGD*** furthur increases the accuracy to about 98%.

4. Adding in Dropout rates in this particular model reduces the overall accuracy as observed that increasing the dropout rate decreases the accuracy.

5. Setting the loss function to ***sparse_categorical_crossentropy*** gives the least losses and highest accuracy.

6. Adding **early Stopping** does not really help in this model as the epoch is already very low and the accurarcy generally has an increasing trend throughout the training process.